server/schedulers/utils.go

```diff
@@ -136,15 +136,170 @@ func adjustBalanceLimit(cluster opt.Cluster, kind core.ResourceKind) uint64 {
+func (s *ScoreInfo) SetScore(score float64) {
+	s.score = score
+}
+
+// ScoreInfos is used for sorting ScoreInfo.
+type ScoreInfos struct {
+	scoreInfos []*ScoreInfo
+	isSorted   bool
+}
+
+// NewScoreInfos returns a ScoreInfos.
+func NewScoreInfos() *ScoreInfos {
+	return &ScoreInfos{
+		scoreInfos: make([]*ScoreInfo, 0),
+	}
+}
+
+// Add adds a scoreInfo into the slice.
+func (s *ScoreInfos) Add(scoreInfo *ScoreInfo) {
+	infosLen := len(s.scoreInfos)
+	if s.isSorted == true && infosLen != 0 && s.scoreInfos[infosLen-1].score > scoreInfo.score {
+		s.isSorted = false
+	}
+	s.scoreInfos = append(s.scoreInfos, scoreInfo)
+}
+
+// Len returns length of slice.
+func (s *ScoreInfos) Len() int { return len(s.scoreInfos) }
+
+// Less returns if one number is less than another.
+func (s *ScoreInfos) Less(i, j int) bool { return s.scoreInfos[i].score < s.scoreInfos[j].score }
+
+// Swap switches out two numbers in slice.
+func (s *ScoreInfos) Swap(i, j int) {
+	s.scoreInfos[i], s.scoreInfos[j] = s.scoreInfos[j], s.scoreInfos[i]
+}
+
+// Sort sorts the slice.
+func (s *ScoreInfos) Sort() {
+	if !s.isSorted {
+		sort.Sort(s)
+		s.isSorted = true
+	}
+}
+
+// GetScoreInfos returns the scoreInfos.
+func (s *ScoreInfos) GetScoreInfos() []*ScoreInfo {
+	return s.scoreInfos
+}
+
+// Min returns the min of the slice.
+func (s *ScoreInfos) Min() *ScoreInfo {
+	s.Sort()
+	return s.scoreInfos[0]
+}
+
+// Mean returns the mean of the slice.
+func (s *ScoreInfos) Mean() float64 {
+	if s.Len() == 0 {
+		return 0
+	}
+
+	var sum float64
+	for _, info := range s.scoreInfos {
+		sum += info.score
+	}
+
+	return sum / float64(s.Len())
+}
+
+// StdDev returns the standard deviation of the slice.
+func (s *ScoreInfos) StdDev() float64 {
+	if s.Len() == 0 {
+		return 0
+	}
+
+	var res float64
+	mean := s.Mean()
+	for _, info := range s.GetScoreInfos() {
+		diff := info.GetScore() - mean
+		res += diff * diff
+	}
+	res /= float64(s.Len())
+	res = math.Sqrt(res)
+
+	return res
+}
+
+// MeanStoresStats returns the mean of stores' stats.
+func MeanStoresStats(storesStats map[uint64]float64) float64 {
+	if len(storesStats) == 0 {
+		return 0.0
+	}
+
+	var sum float64
+	for _, storeStat := range storesStats {
+		sum += storeStat
+	}
+	return sum / float64(len(storesStats))
+}
+
+// NormalizeStoresStats returns the normalized score scoreInfos. Normolize: x_i => (x_i - x_min)/x_avg.
+func NormalizeStoresStats(storesStats map[uint64]float64) *ScoreInfos {
+	scoreInfos := NewScoreInfos()
+
+	for storeID, score := range storesStats {
+		scoreInfos.Add(NewScoreInfo(storeID, score))
+	}
+
+	mean := scoreInfos.Mean()
+	if mean == 0 {
+		return scoreInfos
+	}
+
+	minScore := scoreInfos.Min().GetScore()
+
+	for _, info := range scoreInfos.GetScoreInfos() {
+		info.SetScore((info.GetScore() - minScore) / mean)
+	}
+
+	return scoreInfos
+}
+
+// AggregateScores aggregates stores' scores by using their weights.
+func AggregateScores(storesStats []*ScoreInfos, weights []float64) *ScoreInfos {
```

Comment:
Check `len(storesStats)` and `len(weights)`. Maybe only consider `storesStats[..L]` and `weights[..L]` with `L` is the smaller length.